pep-0696.rst

```diff
@@ -156,44 +157,126 @@ literal "``...``".
+``TypeVarLike`` must be the same type (a ``TypeVar``'s default must be
+a ``TypeVar``, etc.).
 
 `This could be used on builtins.slice <https://github.com/python/typing/issues/159>`__
 where the ``start`` parameter should default to ``int``, ``stop``
 default to the type of ``start`` and step default to ``int | None``.
 
-.. code:: py
+.. code-block:: py
 
    StartT = TypeVar("StartT", default=int)
    StopT = TypeVar("StopT", default=StartT)
    StepT = TypeVar("StepT", default=int | None)
 
-   class slice(Generic[StartT, StopT, StepT]): ...  # Valid
+   class slice(Generic[StartT, StopT, StepT]): ...
+
+   reveal_type(slice())                        # type is slice[int, int, int | None]
+   reveal_type(slice[str]())                   # type is slice[str, str, int | None]
+   reveal_type(slice[str, bool, timedelta]())  # type is slice[str, bool, timedelta]
+
+When using a ``TypeVarLike`` as the default to another ``TypeVarLike``.
+Where ``T1`` is the default for ``T2`` the following rules apply.
+
+.. _scoping-rules:
+
+Scoping Rules
+~~~~~~~~~~~~~
+
+``T1`` must be used before ``T2`` in the parameter list of the generic,
+or be bound in an outer class or function scope.
+
+.. code-block:: py
 
-   reveal_type(slice())                       # type is slice[int, int, int | None]
-   reveal_type(slice[str]())                  # type is slice[str, str, int | None]
-   reveal_type(slice[str, str, timedelta]())  # type is slice[str, str, timedelta]
+   DefaultT = TypeVar("DefaultT", default=T)
 
-   StartT = TypeVar("StartT", default="StopT")
+   class Foo(Generic[T, DefaultT]): ...   # Valid
+   def bar(x: T, y: DefaultT): ...        # Valid
+   class Foo(Generic[T]):
+       class Bar(Generic[DefaultT]): ...  # Valid
+   def outer(x: T):
+       def inner(y: DefaultT): ...        # Valid
+
+   StartT = TypeVar("StartT", default="StopT")  # Swapped defaults around from previous example
    StopT = TypeVar("StopT", default=int)
    class slice(Generic[StartT, StopT, StepT]): ...
-                       ^^^^^^  # Invalid: ordering does not allow StopT to bound yet
+                     # ^^^^^^ Invalid: ordering does not allow StopT to be bound
+   def baz(x: DefaultT, y: T): ...
+            # ^^^^^^^^ Invalid: ordering does not allow DefaultT to be bound
+
+Bound Rules
+~~~~~~~~~~~
+
+``T2``'s bound must be a subtype of ``T1``'s bound.
+
+.. code-block:: py
+
+   T = TypeVar("T", bound=float)
+   TypeVar("Ok", default=T, bound=int)        # Valid
+   TypeVar("AlsoOk", default=T, bound=float)  # Valid
+   TypeVar("Invalid", default=T, bound=str)   # Invalid: str is not a subtype of float
+
+Constraint Rules
+~~~~~~~~~~~~~~~~
+
+Either ``T2``'s constraint must contain the bound of ``T1``, or the
+constraints of ``T2`` must be a superset of the constraints of ``T1``.
+
+.. code-block:: py
+
+   T1 = TypeVar("T1", bound=int)
+   TypeVar("Ok", int, str, default=T1)                # Valid
```

Comment:
This seems sketchy because `T1` may be solved to a subtype of `int`, but `Ok` can only be `int` or `str` exactly.